CP-20411: AMD MxGPU in datamodel and vgpu-type

thomassa opened the pull request:
Extend datamodel.ml and xapi_vgpu_type.ml with the new items
that will be needed for running virtual GPUs on AMD MxGPU cards

thomassa commented:
(Not yet dev-tested.)

mseri commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+module MxGPU = struct
+  let amd_vendor_id = 0x1002
+
+  let ( *** ) = Int64.mul
```
Should we move those `Int64` definition on the top level given that they are shared among all the vgpu modules?

mseri commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+    let size_of_pgpu = (* counterpart of bar_size in gvt-g *)
+      Constants.pgpu_default_size
+    in
+    let rec collect acc = function
```
I am not sure if this is worth doing, given that `Stdext` description claims it is deprecated, however we can simplify this function (and the sister Intel's one) using `Listext.filter_map` (https://github.com/xapi-project/stdext/blob/master/lib/listext.mli#L157 I had not found it when I did the change in the Intel vgpu)

mseri commented on ocaml/idl/datamodel.ml:
```diff
@@ -8633,6 +8633,8 @@ let pci =
       field ~qualifier:StaticRO ~ty:String ~lifecycle:[Published, rel_boston, ""] "pci_id" "PCI ID of the physical device" ~default_value:(Some (VString ""));
       field ~qualifier:DynamicRO ~ty:Int ~lifecycle:[] ~default_value:(Some (VInt 1L)) "functions" "Number of physical + virtual PCI functions" ~internal_only:true;
+      field ~qualifier:DynamicRO ~ty:(Set (Ref _pci)) ~lifecycle:[Published, rel_falcon, ""] "virtual_functions" "Set of VF PCI devices provided by this physical (PF) PCI device";
+      field ~qualifier:StaticRO ~ty:(Ref _pci) ~lifecycle:[Published, rel_falcon, ""] "physical_function" "The PF (physical PCI device) that provides this VF" ~default_value:(Some (VRef null_ref));
```
I think you need to add the new fields to  the `Pci.create` function (https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_pci.ml#L52)

robhoes replied:
Add `~internal_only:true` for both?

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -37,10 +37,17 @@ module Identifier = struct
   }
 
+  type mxgpu_id = {
+    pdev_id : int; (* PCI id of the PF, not the VFs *)
```
Let's explicitly say "device ID", because "PCI ID" is often used for the bus address.

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+  let ( /// ) = Int64.div
+  let ( +++ ) = Int64.add
+  let ( --- ) = Int64.sub
+  let mib x = List.fold_left Int64.mul x [1024L; 1024L]
```
Quite hard-core to use a fold here :) Also, we've just defined `***`, so why not just `x *** 1024L *** 1024L`?

thomassa replied:
This whole module is just a copy of the Intel module with minimal changes.
I wanted to keep the changes as small as possible so that it will be easy to refactor this two modules to be based on a shared functor.
That refactoring will be done in a future pull request.

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -541,21 +687,27 @@ let find_or_create_supported_types ~__context ~pci
-    else [find_or_create ~__context passthrough_gpu]
-  end
+  match vendor_id with
+  | x when x = Nvidia.nvidia_vendor_id -> begin
```
Or simply `| Nvidia.nvidia_vendor_id -> `? Also, you don't need the begin/end here.

thomassa replied:
With `| Nvidia.nvidia_vendor_id` it didn't compile.

robhoes replied:
Oh right, I see why now...

thomassa replied:
Right, it's possible to match against a constant only if it's a literal, which is a bit frustrating in situations like this. I still prefer it to a whole series of `else if` though:
```
if vendor_id = Nvidia.nvidia_vendor_id then begin
    ....
end else if vendor_id = Intel.intel_vendor_id then begin
    ....
end else if vendor_id = MxGPU.amd_vendor_id then begin
    ....
end else begin
   ....
end
```

robhoes replied:
Yes, I agree.

robhoes replied:
Regarding the begin/end, you may need them in if-statements, but not inside match cases (unless you're nesting matches perhaps).

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+
+  let find_or_create_supported_types ~__context ~pci
+      ~is_system_display_device
+      ~is_host_display_enabled
```
Are these final two args appropriate? I believe they are special to the Intel GVT-d/g.

thomassa replied:
I think any PCI device can be hidden.
I'm not sure about the host display. Can a host use one of these cards for its own display?

robhoes replied:
I think it is specifically for Intel graphics due to these being integrated, on-chip GPUs. But check with Paul to be sure.

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+    whitelist
+    |> List.map
+      Identifier.(fun conf ->
+          let vgpus_per_pgpu =
```
Instead of this renaming, can we not just use one name consistently?

thomassa replied:
Again, this is to stay as close as possible to the form of the `Intel` module, for ease of future refactoring where I imagine we will have a function like `get_vgpus_per_pgpu` in the modules we pass to the functor.

robhoes commented on ocaml/xapi/xapi_vgpu_type.ml:
```diff
@@ -533,6 +546,139 @@ module Intel = struct
+            let vgpu_size =
+              size_of_pgpu /// vgpus_per_pgpu
+            in
+            let internal_config = let open Xapi_globs in
```
`open Xapi_globs`?

robhoes commented on ocaml/xapi/xapi_globs.ml:
```diff
@@ -786,6 +786,8 @@ let igd_passthru_vendor_whitelist = ref []
 
 let gvt_g_whitelist = ref "/etc/gvt-g-whitelist"
 
+let mxgpu_whitelist = ref "/etc/mxgpu-whitelist"
```
You should add a config file option as well, below.

robhoes commented:
It doesn't actually compile yet...
```
File "xapi_xenops.ml", line 638, characters 11-306:                                          
Warning 8: this pattern-matching is not exhaustive.======================    ] 02130 / 02208
Here is an example of a value that is not matched:
`mxgpu
```

robhoes commented on ocaml/xapi/xapi_xenops.ml:
```diff
@@ -642,6 +645,8 @@ module MD = struct
            | `gvt_g ->
              (of_gvt_g_vgpu ~__context vm vgpu_record) :: acc)
+           | `mxgpu ->
+             (of_mxgpu_vgpu ~__context vm vgpu_record) :: acc)
```
I think the () are wrong... (made the same mistake myself).

robhoes commented:
Also `make test` will fail... :(

thomassa commented:
I think this can be merged now:
I've addressed all the comments, except for issues that will go away in the refactoring/cleanup of the `Intel` and `MxGPU` modules in `Xapi_vgpu_type`.
(This first pullreq is just laying groundwork.)

mseri commented:
Fine with me

thomassa commented:
I have already added the question about `is_system_display_device && (is_host_display_enabled || not is_pci_hidden)` into task CP-20456.
